src/section/section_chain.rs

```diff
@@ -0,0 +1,1114 @@
+// then converting it into `SectionChain` using `try_from` which fails when the chain is invalid.
+// This makes it impossible to obtain invalid `SectionChain` from malformed serialized data, thus
+// making `SectionChain` "correct by deserialization".
+#[derive(Deserialize)]
+#[serde(rename = "SectionChain")]
+struct Deserialized {
+    root: bls::PublicKey,
+    tree: Vec<Block>,
+}
+
+impl TryFrom<Deserialized> for SectionChain {
+    type Error = IntegrityError;
+
+    fn try_from(src: Deserialized) -> Result<Self, Self::Error> {
+        let mut prev_block: Option<&Block> = None;
+
+        for (tree_index, block) in src.tree.iter().enumerate() {
+            let parent_key = if block.parent_index == 0 {
+                &src.root
+            } else if block.parent_index > tree_index {
+                return Err(IntegrityError::WrongBlockOrder);
+            } else if let Some(key) = src.tree.get(block.parent_index - 1).map(|block| &block.key) {
+                key
+            } else {
+                return Err(IntegrityError::ParentNotFound);
+            };
+
+            if !block.verify(parent_key) {
+                // Invalid signature
+                return Err(IntegrityError::FailedSignature);
+            }
+
+            if let Some(prev_block) = prev_block {
+                if block.parent_index < prev_block.parent_index {
+                    // Wrong order of block that have neither parent-child, not sibling relation.
+                    return Err(IntegrityError::WrongBlockOrder);
+                }
+
+                if block.parent_index == prev_block.parent_index && block <= prev_block {
+                    // Wrong sibling order
+                    return Err(IntegrityError::WrongBlockOrder);
+                }
+            }
+
+            prev_block = Some(block);
+        }
+
+        Ok(Self {
+            root: src.root,
+            tree: src.tree,
+        })
+    }
+}
+
+#[derive(Debug, Error, Eq, PartialEq)]
+enum IntegrityError {
+    #[error("signature check failed")]
+    FailedSignature,
+    #[error("parent key not found in the chain")]
+    ParentNotFound,
+    #[error("chain blocks are in a wrong order")]
+    WrongBlockOrder,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn insert_last() {
+        let mut expected_keys = vec![];
+        let (mut last_sk, pk) = gen_keypair();
+
+        let mut chain = SectionChain::new(pk);
+        expected_keys.push(pk);
+
+        for _ in 0..10 {
+            let last_pk = &expected_keys[expected_keys.len() - 1];
+            let (sk, pk, sig) = gen_signed_keypair(&last_sk);
+
+            assert_eq!(chain.insert(last_pk, pk, sig), Ok(()));
+
+            expected_keys.push(pk);
+            last_sk = sk;
+        }
+
+        assert_eq!(chain.keys().copied().collect::<Vec<_>>(), expected_keys);
+    }
+
+    #[test]
+    fn insert_fork() {
```

Comment:
what if the fork is 
first `0->1->2` then insert `0->3` ?
will the chain becomes `0 , 3 , 1 , 2` (when 1>3) ,   `0, 1, 2 , 3` (when 1<=3) or  `0, 1 , 3 , 2` (when 1<=3 AND 2 > 3) ?

Reply:
If `1 > 3` then the chain will be `0, 3, 1, 2`. If `1 < 3` then it will be `0, 1, 3, 2`. It will never be `0, 1, 2, 3`.

Reply:
will it be worth to have a test, or just expand the current one, to confirm this behaviour ?

Reply:
also, I am wondering the case that if there is `3->4` to be inserted later on.
if the chain was `0 , 3 , 1 , 2`, it becomes  `0 , 3 , 4 , 1 , 2` and truncate(2) gives `1, 2`
when the chain was  `0 , 1 , 3 , 2`, it becomes  `0 , 1 , 3 , 4 , 2` and truncate(2) shall still give `1, 2`
right?

Reply:
That's right, because `truncate` only follows the parents and ignores the siblings.